### PR TITLE
Generate transient storage layout

### PR DESCRIPTION
**Pull request by matheusaaguiar**

### CODE REVIEW

File: libsolidity/interface/StorageLayout.cpp [hunk 1]
```diff
@@ -20,10 +20,11 @@
 
 #include <libsolidity/ast/TypeProvider.h>
 
+using namespace std::literals;
 using namespace solidity;
 using namespace solidity::frontend;
 
-Json StorageLayout::generate(ContractDefinition const& _contractDef)
+Json StorageLayout::generate(ContractDefinition const& _contractDef, DataLocation const _location)
```
Comment (matheusaaguiar) [inline libsolidity/interface/StorageLayout.cpp:27]: The changes here are really just about filtering the variables by their data location, so I didn't create a separate class. But maybe this one could be renamed to not be confused with the storage data location? I haven't thought too much about it, but I guess something like `StateLayout` or `StateVariablesLayout`...

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -2140,12 +2140,21 @@ FunctionType const* ContractType::newExpressionType() const
 	return m_constructorType;
 }
 
-std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables() const
+std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(DataLocation _location) const
 {
+	VariableDeclaration::Location location = [&]() {
+		switch (_location)
+		{
+			case DataLocation::Storage: return VariableDeclaration::Location::Unspecified; // By default state variables have unspecified (storage) data location
```
Comment (matheusaaguiar) [inline libsolidity/ast/Types.cpp]: I think this should be refactored in a separate PR so variables at storage are explicitly assigned `VariableDeclaration::Location::Storage` (note this is a different enum from `DataLocation` which is in `Types.h`).

File: Changelog.md [hunk 3]
```diff
@@ -7,10 +7,12 @@ Language Features:
 
 
 Compiler Features:
+ * Commandline Interface: Output transient storage layout with option `--transient-storage-layout`.
```
Comment (matheusaaguiar) [inline Changelog.md]:
```suggestion
 * Commandline Interface: Output transient storage layout with option ``--transient-storage-layout``.
```

File: Changelog.md [hunk 4]
```diff
@@ -7,10 +7,12 @@ Language Features:
 
 
 Compiler Features:
+ * Commandline Interface: Output transient storage layout with option ``--transient-storage-layout``.
  * Error Reporting: Unimplemented features are now properly reported as errors instead of being handled as if they were bugs.
  * EVM: Support for the EVM version "Prague".
  * SMTChecker: Add CHC engine check for underflow and overflow in unary minus operation.
  * SMTChecker: Replace CVC4 as a possible BMC backend with cvc5.
+ * Standard JSON Interface: Output the storage layout of a contract when artifact ``transientStorageLayout`` is requested.
```
Comment (nikola-matic) [inline Changelog.md]:
```suggestion
 * Standard JSON Interface: Output the transient storage layout of a contract when artifact ``transientStorageLayout`` is requested.
```

File: libsolidity/interface/StorageLayout.cpp (hunk 1 above)
Comment (nikola-matic) [reply libsolidity/interface/StorageLayout.cpp:27]: This is what I spent the significant portion of my time thinking about before I left for vacation, because `StorageLayout` does everything we need it to do for transient storage as well, since we decided that it should be a 1:1 copy of regular storage layout anyway. Now that you're passing data location here, we can add unimplemented asserts where necessary (i.e. no transient layout for reference types), and leave it at that.

I don't really see the value in refactoring this, at least not at the moment, so leave it as is.

File: libsolidity/ast/Types.cpp [hunk 5]
```diff
@@ -2140,12 +2140,21 @@ FunctionType const* ContractType::newExpressionType() const
...
+			case DataLocation::Transient: return VariableDeclaration::Location::Transient;
+			default: solAssert(false);
+		}
+	}();
+
 	std::vector<VariableDeclaration const*> variables;
 	for (ContractDefinition const* contract: m_contract.annotation().linearizedBaseContracts | ranges::views::reverse)
 		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!(variable->isConstant() || variable->immutable()))
+			if (!(variable->isConstant() || variable->immutable()) && variable->referenceLocation() == location)
```
Comment (nikola-matic) [inline libsolidity/ast/Types.cpp]: Is this check necessary - we're iterating over state variables here, and these can't be memory or calldata anyway, which only leaves unspecified/transient? I would assume that it is since you've added it here, but am not sure why.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 6]
```diff
@@ -101,6 +103,16 @@ std::string IRGenerator::generate(
 	std::map<ContractDefinition const*, std::string_view const> const& _otherYulSources
 )
 {
+	auto static notTransient = [](VariableDeclaration const* _varDeclaration) {
```
Comment (nikola-matic) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Why is this static?

File: test/cmdlineTests/transient_storage_layout_smoke/output.json
Comment (nikola-matic) [inline test/cmdlineTests/transient_storage_layout_smoke/output.json:1]: I'd rename this to `...smoke_empty`.

File: libsolidity/ast/Types.cpp (hunk 5 above)
Comment (matheusaaguiar) [reply libsolidity/ast/Types.cpp]: Constants and immutables can also be unspecified, and they are not stored anywhere.
If we refactor as suggested [here](https://github.com/ethereum/solidity/pull/15247#discussion_r1671291082), then I think we can remove the check.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 6 above)
Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGenerator.cpp]: I don't know why I made this static. Let's just remove it. :)

File: libsolidity/interface/StorageLayout.cpp [hunk 7]
```diff
@@ -34,19 +35,27 @@ Json StorageLayout::generate(ContractDefinition const& _contractDef)
 	auto contractType = dynamic_cast<ContractType const*>(typeType->actualType());
 	solAssert(contractType, "");
 
+	std::string locationKey = [&]() {
+		switch (_location)
+		{
+			case DataLocation::Storage: return "storage"s;
+			case DataLocation::Transient: return "transient-storage"s;
+			default: solAssert(false);
```
Comment (nikola-matic) [inline libsolidity/interface/StorageLayout.cpp]:
```suggestion
		case DataLocation::Storage: return "storage"s;
		case DataLocation::Transient: return "transient-storage"s;
		default: solAssert(false);
```

File: libsolidity/ast/Types.cpp [hunk 8]
```diff
@@ -2140,12 +2140,21 @@ FunctionType const* ContractType::newExpressionType() const
...
 
-std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables() const
+std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(DataLocation _location) const
 {
+	VariableDeclaration::Location location = [&]() {
+		switch (_location)
+		{
+			case DataLocation::Storage: return VariableDeclaration::Location::Unspecified; // By default state variables have unspecified (storage) data location
+			case DataLocation::Transient: return VariableDeclaration::Location::Transient;
+			default: solAssert(false);
```
Comment (nikola-matic) [inline libsolidity/ast/Types.cpp]:
```suggestion
		case DataLocation::Storage: return VariableDeclaration::Location::Unspecified; // By default state variables have unspecified (storage) data location
		case DataLocation::Transient: return VariableDeclaration::Location::Transient;
		default: solAssert(false);
```

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 9]
```diff
@@ -1103,8 +1115,9 @@ void IRGenerator::resetContext(ContractDefinition const& _contract, ExecutionCon
 	m_context = std::move(newContext);
 
 	m_context.setMostDerivedContract(_contract);
-	for (auto const& var: ContractType(_contract).stateVariables())
-		m_context.addStateVariable(*std::get<0>(var), std::get<1>(var), std::get<2>(var));
+	for (auto const& location: {DataLocation::Storage, DataLocation::Transient})
```
Comment (nikola-matic) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Might make sense to make the required data location `std::optional` and set to `std::nullopt` by default - this would alleviate then need to change both this section of code as well as the one in `ContractCompiler`.

File: libsolidity/ast/Types.cpp (hunk 2 above)
Comment (nikola-matic) [reply libsolidity/ast/Types.cpp]: Why do we even need two enums for this? It makes it significantly more annoying in the first place. Adding an `Unspecified` option to the `DataLocation` enum would be enough; plus it's not a class member enum so it can be aliased as well (unlike `VariableDeclaration::Location`). But yeah, a problem for another time.

File: test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{
+	"language": "Solidity",
+	"sources": {
+		"fileA": {
+			"content": "//SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\ncontract A { }"
+		},
+		"fileB": {
+			"content": "//SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\ncontract A { uint transient x; uint transient y; }"
```
Comment (nikola-matic) [inline test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json]: Are both contract `A` on purpose?

File: test/cmdlineTests/transient_storage_layout_value_types/in.sol [hunk 11]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.0;
+contract A {
+    uint transient x;
+    int transient y;
+    address transient addr;
+    bool transient b;
+}
```
Comment (nikola-matic) [inline test/cmdlineTests/transient_storage_layout_value_types/in.sol]:
```suggestion
}

```

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/in.sol [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+contract A {
+    uint transient x;
+    uint y;
+    int transient z;
+    int w;
+    address transient addr;
+    address d;
+    bool transient b;
+    bool c;
+}
```
Comment (nikola-matic) [inline test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/in.sol]:
```suggestion
}

```

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json [hunk 13]
```diff
@@ -0,0 +1,13 @@
+{
+	"language": "Solidity",
+	"sources": {
+		"fileA": {"urls": ["transient_storage_layout_value_types_interleaved_with_storage/in.sol"]}
+	},
+	"settings": {
+		"outputSelection": {
+			"fileA": {
+				"A": ["transientStorageLayout"]
```
Comment (nikola-matic) [inline test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json]: Request the `storageLayout` artifact here as well.

File: libsolidity/ast/Types.cpp (hunk 5 above)
Comment (nikola-matic) [reply libsolidity/ast/Types.cpp]: Constants and immutables are excluded in the first conditional `!(variable->isConstant() || variable->immutable())`, so that's not an issue. The question on my side was bad though - the reason you have this check is to filter storage vs transient storage state variables when generating layouts, so this check is necessary regardless.

File: solc/CommandLineParser.cpp [hunk 14]
```diff
@@ -759,7 +759,8 @@ General Information)").c_str(),
 		(CompilerOutputs::componentName(&CompilerOutputs::natspecUser).c_str(), "Natspec user documentation of all contracts.")
 		(CompilerOutputs::componentName(&CompilerOutputs::natspecDev).c_str(), "Natspec developer documentation of all contracts.")
 		(CompilerOutputs::componentName(&CompilerOutputs::metadata).c_str(), "Combined Metadata JSON whose IPFS hash is stored on-chain.")
-		(CompilerOutputs::componentName(&CompilerOutputs::storageLayout).c_str(), "Slots, offsets and types of the contract's state variables.")
+		(CompilerOutputs::componentName(&CompilerOutputs::storageLayout).c_str(), "Slots, offsets and types of the contract's state variables located at storage.")
+		(CompilerOutputs::componentName(&CompilerOutputs::transientStorageLayout).c_str(), "Slots, offsets and types of the contract's state variables located at transient storage.")
```
Comment (nikola-matic) [inline solc/CommandLineParser.cpp]:
```suggestion
		(CompilerOutputs::componentName(&CompilerOutputs::storageLayout).c_str(), "Slots, offsets and types of the contract's state variables located in storage.")
		(CompilerOutputs::componentName(&CompilerOutputs::transientStorageLayout).c_str(), "Slots, offsets and types of the contract's state variables located in transient storage.")
```

File: test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json (hunk 10 above)
Comment (matheusaaguiar) [reply test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json]: No, that was just copy&paste :smile: 
Fixed.

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json (hunk 13 above)
Comment (matheusaaguiar) [reply test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json]: Done.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 9 above)
Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGenerator.cpp]: Yeah, I thought about that at first but expected it to be more complicated than what I have actually come up with.

File: Changelog.md [hunk 15]
```diff
@@ -8,11 +8,13 @@ Language Features:
 
 Compiler Features:
  * Command Line Interface: Do not perform IR optimization when only unoptimized IR is requested.
+ * Commandline Interface: Output transient storage layout with option ``--transient-storage-layout``.
  * Error Reporting: Unimplemented features are now properly reported as errors instead of being handled as if they were bugs.
  * EVM: Support for the EVM version "Prague".
  * SMTChecker: Add CHC engine check for underflow and overflow in unary minus operation.
  * SMTChecker: Replace CVC4 as a possible BMC backend with cvc5.
  * Standard JSON Interface: Do not perform IR optimization when only unoptimized IR is requested.
+ * Standard JSON Interface: Output the transient storage layout of a contract when artifact ``transientStorageLayout`` is requested.
```
Comment (nikola-matic) [inline Changelog.md]: Can't believe I forgot this - but you're gonna have to add this to the docs :)

**Review by nikola-matic — DISMISSED** [review]
@cameel please take over.

File: Changelog.md [hunk 16]
```diff
@@ -8,11 +8,13 @@ Language Features:
 
 Compiler Features:
  * Command Line Interface: Do not perform IR optimization when only unoptimized IR is requested.
+ * Commandline Interface: Output transient storage layout with option ``--transient-storage-layout``.
```
Comment (cameel) [inline Changelog.md]: Looks like we don't have any test coverage for the CLI variant in the PR, do we?

File: test/cmdlineTests/standard_outputs_on_compilation_error/output.json [hunk 17]
```diff
@@ -37,6 +37,10 @@
                     "storage": [],
                     "types": null
                 },
+                "transientStorageLayout": {
+                    "transient-storage": [],
```
Comment (cameel) [inline test/cmdlineTests/standard_outputs_on_compilation_error/output.json]:
```suggestion
                    "transientStorage": [],
```

File: test/cmdlineTests/transient_storage_layout_smoke_empty/input.json [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+	"sources": {
+		"fileA": {
+			"content": "//SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\ncontract A { }"
+		}
+	},
+	"settings": {
+		"outputSelection": {
+			"fileA": {
+				"A": [ "transientStorageLayout" ],
+				"": [ "transientStorageLayout" ]
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_smoke_empty/input.json]: We don't have file-level transient storage layout output, so requesting it is pointless here (unless you're specifically testing that it does nothing, but that does not seem very useful).

It's also done in some of the other tests.

Comment (cameel) [reply test/cmdlineTests/transient_storage_layout_smoke_empty/input.json]:
```suggestion
				"A": ["transientStorageLayout"],
				"": ["transientStorageLayout"]
```

File: test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+			"content": "//SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\ncontract A { }"
+		},
+		"fileB": {
+			"content": "//SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\ncontract B { uint transient x; uint transient y; }"
+		}
+	},
+	"settings": {
+		"outputSelection": {
+			"fileA": {
+				"A": [ "transientStorageLayout" ],
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json]: You forgot to request it for contract `B`, so the layout does not show up in the output.

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/in.sol (hunk 12 above)
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/in.sol]: I'd also add some constants and immutables here to make sure they're not included in either layout.

Also different visibilities (`public`, `private`, etc.).

Could also be separate tests, but IMO it would be fine to keep it in one test since CLI tests are pretty verbose.

File: test/cmdlineTests/transient_storage_layout_value_types/in.sol (hunk 11 above)
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_value_types/in.sol]: Can you add some more types? e.g. UDVT, function pointer, enum, contract.

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json [hunk 20]
```diff
@@ -0,0 +1,13 @@
+{
+	"language": "Solidity",
+	"sources": {
+		"fileA": {"urls": ["transient_storage_layout_value_types_interleaved_with_storage/in.sol"]}
+	},
+	"settings": {
+		"outputSelection": {
+			"fileA": {
+				"A": ["storageLayout","transientStorageLayout"]
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json]: It would probably make sense to always request both in these tests (except for the ones that specifically test that `transientStorageLayout` triggers the right output), so that we can see what ends up in each one.

I'd probably even go as far as to add this output to existing `storageLayout` tests to get an easy check that in all of them there's no transient storage detected.

Comment (cameel) [reply test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/input.json]: Also:
```suggestion
				"A": ["storageLayout", "transientStorageLayout"]
```

**Review by cameel — COMMENTED** [review]
Here are a few initial remarks after going through tests.

So far only minor stuff, other than the fact that we haven't even one test covering CLI output. And that docs have not been updated.

File: libsolidity/interface/StorageLayout.cpp [hunk 21]
```diff
@@ -34,19 +35,27 @@ Json StorageLayout::generate(ContractDefinition const& _contractDef)
...
+		}
+	}();
 	Json variables = Json::array();
-	for (auto [var, slot, offset]: contractType->stateVariables())
+	for (auto [var, slot, offset]: contractType->stateVariables(_location))
 		variables.emplace_back(generate(*var, slot, offset));
 
 	Json layout;
-	layout["storage"] = std::move(variables);
+	layout[locationKey] = std::move(variables);
```
Comment (cameel) [inline libsolidity/interface/StorageLayout.cpp]: I don't think we even need to change this key to match the location. `"storage"` works for both transient and normal storage and makes things simpler, because the structure stays the same.

Actually it would have been better off named something like `"variables"` but we can't change that now.

File: libsolidity/ast/Types.cpp [hunk 22]
```diff
@@ -2140,12 +2140,26 @@ FunctionType const* ContractType::newExpressionType() const
 	return m_constructorType;
 }
 
-std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables() const
+std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(std::optional<DataLocation> _location) const
 {
+	auto location = [&]() -> std::optional<VariableDeclaration::Location> {
```
Comment (cameel) [inline libsolidity/ast/Types.cpp]: I really don't like how ambiguous this style of declaration is. It looks like you're declaring a lambda until the very last moment when you realize there's `()` at the end and it's actually the result of that lambda.

If you really have to have a lambda here, I'd much rather have this:
```suggestion
	std::optional<VariableDeclaration::Location> location = [&]() {
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 23]
```diff
@@ -478,13 +478,6 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	Type const* varType = _variable.annotation().type;
 	solAssert(!!varType, "Variable type not provided.");
 
-	if (_variable.referenceLocation() == VariableDeclaration::Location::Transient)
-		m_errorReporter.unimplementedFeatureError(
-			6715_error,
-			_variable.location(),
-			"Transient storage is not yet implemented."
-		);
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp:486]: Now that we can generate the AST again, we should add back the test for it.

File: Changelog.md (hunk 15 above)
Comment (cameel) [inline Changelog.md]: How about this? More compact and we generally call these "outputs" rather than "artifacts" (at least when referring to the the option/JSON key itself).

```suggestion
 * Commandline Interface: Add ``--transient-storage-layout`` output.
 * Error Reporting: Unimplemented features are now properly reported as errors instead of being handled as if they were bugs.
 * EVM: Support for the EVM version "Prague".
 * SMTChecker: Add CHC engine check for underflow and overflow in unary minus operation.
 * SMTChecker: Replace CVC4 as a possible BMC backend with cvc5.
 * Standard JSON Interface: Do not perform IR optimization when only unoptimized IR is requested.
 * Standard JSON Interface: Add ``--transientStorageLayout`` output.
```

**Review by cameel — COMMENTED** [review]
I went through it all now and did not find anything broken. Just a few more testing/styling/wording remarks.

So overall this is mostly fine, just needs docs, CLI coverage, some small tweaks and then we can merge.

File: libsolidity/ast/Types.cpp (hunk 2 above)
Comment (cameel) [reply libsolidity/ast/Types.cpp]: > Why do we even need two enums for this?

`DataLocation` specifies where the data is stored while `VariableDeclaration::Location`  indicates which modifier is on the variable.

> I think this should be refactored in a separate PR so variables at storage are explicitly assigned `VariableDeclaration::Location::Storage`

`Unspecified` does not necessarily mean `Storage`. That's the case only for state variables. And even then it may also represent a constant or an immutable.

I think changing the way it works would just make things confusing in a different way. Then we'd be asking why it doesn't reflect what's on the variable :) The current usage does have some logic behind it. This is just another case of things being too vaguely named, omitting important context. Maybe something like `DeclaredLocation` or `LocationModifier` would be clearer.

> Adding an `Unspecified` option to the `DataLocation` enum would be enough;

It would be misleading because this `Unspecified` would have a different meaning than the one in `VariableDeclaration::Location` and it still would not be a 1:1 translation, IMO `nullopt` is a better choice.

File: libsolidity/ast/Types.cpp (hunk 22 above)
Comment (matheusaaguiar) [reply libsolidity/ast/Types.cpp]: Yeah, I agree that this is misleading until the last moment (and if you catch it) when you see the `()`.
I will add the explicit type, I used `auto` mainly because I have to have the explicit `-> std::optional<VariableDeclaration::Location>` otherwise, the `std::nullopt` is deduced as the return type and conflicts with the actual desired return type.

File: Changelog.md [hunk 24]
```diff
@@ -8,11 +8,13 @@ Language Features:
 
 Compiler Features:
  * Command Line Interface: Do not perform IR optimization when only unoptimized IR is requested.
+ * Commandline Interface: Add ``--transient-storage-layout`` output.
  * Error Reporting: Unimplemented features are now properly reported as errors instead of being handled as if they were bugs.
  * EVM: Support for the EVM version "Prague".
  * SMTChecker: Add CHC engine check for underflow and overflow in unary minus operation.
  * SMTChecker: Replace CVC4 as a possible BMC backend with cvc5.
  * Standard JSON Interface: Do not perform IR optimization when only unoptimized IR is requested.
+ * Standard JSON Interface: Add ``--transientStorageLayout`` output.
```
Comment (nikola-matic) [inline Changelog.md]:
```suggestion
 * Standard JSON Interface: Add ``transientStorageLayout`` output.
```

**Review by nikola-matic — COMMENTED** [review]
You haven't added a cmdline test for `--transient-storage-layout` yet, right?

**Comment by matheusaaguiar** [comment]
> You haven't added a cmdline test for `--transient-storage-layout` yet, right?

No I haven't. I wrote a test case in `solc/CommandLineInterface.cpp` but the output contains the path to a temporary file which changes everytime and I am not sure how to handle it. See https://github.com/ethereum/solidity/pull/15247/commits/6b6407c57c7f5e61b253eea47507602216c9cc6e.

**Comment by nikola-matic** [comment]
> > You haven't added a cmdline test for `--transient-storage-layout` yet, right?
> 
> No I haven't. I wrote a test case in `solc/CommandLineInterface.cpp` but the output contains the path to a temporary file which changes everytime and I am not sure how to handle it. See [6b6407c](https://github.com/ethereum/solidity/commit/6b6407c57c7f5e61b253eea47507602216c9cc6e).

I meant something like this - https://github.com/ethereum/solidity/tree/develop/test/cmdlineTests/ir_subobject_order

**Comment by matheusaaguiar** [comment]
> I meant something like this - https://github.com/ethereum/solidity/tree/develop/test/cmdlineTests/ir_subobject_order

Duh, yeah, I forgot we can do that :grin: 
Added.

File: libsolidity/analysis/TypeChecker.cpp (hunk 23 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/TypeChecker.cpp:486]: I added a `astPropertyTest`, I guess this was the test you were referring to, right?

**Comment by matheusaaguiar** [comment]
We have this [section](https://docs.soliditylang.org/en/v0.8.26/internals/layout_in_storage.html#layout-of-state-variables-in-storage) in the docs describing the storage layout and it mentions `storage` very frequently. I was thinking of a good way to include the information about transient storage. I think that a note explaining that **regarding** the layout of state variables, and only that, the same rules apply for both storage and transient storage.
Is that a good approach?

File: docs/internals/layout_in_storage.rst [hunk 25]
```diff
@@ -6,6 +6,9 @@ Layout of State Variables in Storage
 
 .. _storage-inplace-encoding:
 
+.. note::
+     The rules described in this section apply for both storage and transient storage data locations.
```
Comment (cameel) [inline docs/internals/layout_in_storage.rst]: I'd still add a few things:
- Maybe change title to "Layout of State Variables in Storage and Transient Storage"?
- Say that the layouts are completely independent, i.e. storage variables have zero effect on locations of transient storage variables and vice-versa. They can be safely interleaved.
- I'd mention that only value types are supported for transient storage.
- There's a note saying that we consider storage layout a part of the external interface. Is that still true for transient storage? If so, we should say that explicitly.
- Add an index entry for `transient storage`, next to `storage` above.

Comment (cameel) [reply docs/internals/layout_in_storage.rst]: Also, there's a note way below, saying:

> The JSON output format of a contract's storage layout is still considered experimental and is subject to change in non-breaking releases of Solidity.

I don't think it's true any more. When we once broke the layout accidentally in 0.8.x cycle, we considered it very breaking and released a new version a few days later to correct it.

Given that, I wonder if we should now say that about transient storage or is it automatically considered stable by the virtue of being tied to storage layout?

File: test/cmdlineTests/transient_storage_layout/args [hunk 26]
```diff
@@ -0,0 +1 @@
+--transient-storage-layout --pretty-json
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout/args]:
```suggestion
--transient-storage-layout --pretty-json --json-indent 4
```

File: test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+		"fileB": {
+			"content": "//SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\ncontract B { uint transient x; uint transient y; }"
+		}
+	},
+	"settings": {
+		"outputSelection": {
+			"fileA": {
+				"A": ["transientStorageLayout"],
+				"B": ["transientStorageLayout"]
+			}
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json]: `B`'s storage layout is missing from the output. Note that it's defined in `fileB`, not `fileA` :)

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/output.json [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+            "component": "general",
+            "errorCode": "2314",
+            "formattedMessage": "ParserError: Expected identifier but got 'external'
+ --> fileA:8:13:
+  |
+8 |     address external transient addr;
+  |             ^^^^^^^^
+
+",
+            "message": "Expected identifier but got 'external'",
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/output.json]: The test doesn't compile. I guess `external` is not allowed on storage variables?

File: test/cmdlineTests/transient_storage_layout_value_types_packed/in.sol [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+    uint128 transient y;
+    uint128 transient z;
+    address transient addr;
+    bool transient b;
+    MyInt transient w;
+    C transient c;
+    MyEnum transient e;
+    function() external transient p;
+}
+}
```
Comment (cameel) [inline test/cmdlineTests/transient_storage_layout_value_types_packed/in.sol]: The test doesn't compile because of this:
```suggestion
}
```

**Review by cameel — COMMENTED** [review]
A couple of tests have syntax errors and we could use a tiny bit more detail in the docs.

I didn't find anything more serious and these are trivial to fix so you can consider it soft-approved.

**Comment by cameel** [comment]
> I think that a note explaining that **regarding** the layout of state variables, and only that, the same rules apply for both storage and transient storage.
> Is that a good approach?

I generally agree in that we don't have to create a separate section and just repeat everything verbatim, but I think we could use a tiny bit more mentions than just a single note :) See https://github.com/ethereum/solidity/pull/15247#discussion_r1700636368

File: libsolidity/ast/Types.cpp (hunk 22 above)
Comment (cameel) [reply libsolidity/ast/Types.cpp]: I mean I'd just drop the lambda :)

Bit anyway, the way it is now is much better.

Comment (matheusaaguiar) [reply libsolidity/ast/Types.cpp]: Yeah, you are right, after re-writing it, it is not adding anything. I've dropped it.

File: test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json (hunk 27 above)
Comment (matheusaaguiar) [reply test/cmdlineTests/transient_storage_layout_smoke_two_contracts/input.json]: Fixed.

File: test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/output.json (hunk 28 above)
Comment (matheusaaguiar) [reply test/cmdlineTests/transient_storage_layout_value_types_interleaved_with_storage/output.json]: Yeah, it is not allowed. I should have tested before and then actually looked at the result.
Fixed.

File: docs/internals/layout_in_storage.rst (hunk 25 above)
Comment (matheusaaguiar) [reply docs/internals/layout_in_storage.rst]: I have applied all suggestions, but not sure about the notes regarding external interface. 
What do you think @ekpyron?

**Review by cameel — APPROVED** [review]

File: docs/internals/layout_in_storage.rst (hunk 25 above)
Comment (cameel) [reply docs/internals/layout_in_storage.rst]: TBH this is already wrong on `develop` so it's fine to merge the PR regardless and then fix it separately in a smaller doc-only PR.

**Comment by nikola-matic** [comment]
Gonna need to squash the commits - still got some fixups in there.

**Review by nikola-matic — APPROVED** [review]
